Add m2m100

patil-suraj opened the pull request:
# What does this PR do?

Adds the M2M100 model
https://github.com/pytorch/fairseq/tree/master/examples/m2m_100

Fixes #8054

patil-suraj committed:
m2m_100

patil-suraj committed:
no layernorm_embedding

patil-suraj committed:
sinusoidal positional embeddings

patil-suraj committed:
update pos embeddings

patil-suraj committed:
add default config values

patil-suraj committed:
tokenizer

patil-suraj committed:
add conversion script

patil-suraj committed:
fix config

patil-suraj committed:
fix pos embed

patil-suraj committed:
remove _float_tensor

patil-suraj committed:
update tokenizer

patil-suraj committed:
update lang codes

patil-suraj committed:
handle lang codes

patil-suraj committed:
fix pos embeds

patil-suraj committed:
fix spm key

patil-suraj committed:
put embedding weights on device

patil-suraj committed:
remove qa and seq classification heads

patil-suraj committed:
fix convert script

patil-suraj committed:
lang codes pn one line

patil-suraj committed:
fix embeds

patil-suraj committed:
fix tokenizer

patil-suraj committed:
fix tokenizer

patil-suraj committed:
add fast tokenizer

patil-suraj committed:
style

patil-suraj committed:
M2M100MT => M2M100

patil-suraj committed:
fix copyright, style

patil-suraj committed:
tokenizer converter

patil-suraj committed:
vocab file

patil-suraj committed:
remove fast tokenizer

patil-suraj committed:
fix embeds

patil-suraj committed:
fix tokenizer

patil-suraj committed:
fix tests

patil-suraj committed:
add tokenizer tests

patil-suraj committed:
add integration test

patil-suraj committed:
quality

patil-suraj committed:
fix model name

patil-suraj committed:
fix test

patil-suraj committed:
doc

patil-suraj committed:
doc

patil-suraj commented on docs/source/model_doc/m2m_100.rst:
```diff
@@ -0,0 +1,127 @@
+    tokenizer.src_lang = "hi"
+    encoded_hi = tokenizer(hi_text, return_tensors="pt")
+    generated_tokens = model.generate(**encoded_hi, forced_bos_token_id=tokenizer.get_lang_id("fr"))
+    tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)
+    # => "La vie est comme une boîte de chocolat."
```
This one got it right  @lhoestq 😉

lhoestq replied:
Nice ! 🍫🍫🍫

patil-suraj committed:
fix doc

patrickvonplaten commented on docs/source/model_doc/m2m_100.rst:
```diff
@@ -0,0 +1,127 @@
+Training and Generation
+_______________________________________________________________________________________________________________________
+
+M2M100 is a multilingual encoder-decoder (seq-to-seq) model primarily intended for translation task. As the model is
```
```suggestion
M2M100 is a multilingual encoder-decoder (seq-to-seq) model primarily intended for translation tasks. As the model is
```

patrickvonplaten commented on docs/source/model_doc/m2m_100.rst:
```diff
@@ -0,0 +1,127 @@
+_______________________________________________________________________________________________________________________
+
+M2M100 is a multilingual encoder-decoder (seq-to-seq) model primarily intended for translation task. As the model is
+multilingual it expects the sequences in a different format. A special language id token is used as prefix in both the
```
```suggestion
multilingual it expects the sequences in a certain format. A special language id token is used as prefix in both the
```

patrickvonplaten commented on docs/source/model_doc/m2m_100.rst:
```diff
@@ -0,0 +1,127 @@
+
+M2M100 is a multilingual encoder-decoder (seq-to-seq) model primarily intended for translation task. As the model is
+multilingual it expects the sequences in a different format. A special language id token is used as prefix in both the
+source and target text. The source text format is is :obj:`[lang_code] X [eos]`, where :obj:`lang_code` is source
```
```suggestion
source and target text. The source text format is :obj:`[lang_code] X [eos]`, where :obj:`lang_code` is source
```

patrickvonplaten commented on docs/source/model_doc/m2m_100.rst:
```diff
@@ -0,0 +1,127 @@
+M2M100 is a multilingual encoder-decoder (seq-to-seq) model primarily intended for translation task. As the model is
+multilingual it expects the sequences in a different format. A special language id token is used as prefix in both the
+source and target text. The source text format is is :obj:`[lang_code] X [eos]`, where :obj:`lang_code` is source
+language id for source text and target language id for target text, with :obj:`X` being the source or target text
```
```suggestion
language id for both source and target text, with :obj:`X` being the source or target text
```

patrickvonplaten commented on docs/source/model_doc/m2m_100.rst:
```diff
@@ -0,0 +1,127 @@
+multilingual it expects the sequences in a different format. A special language id token is used as prefix in both the
+source and target text. The source text format is is :obj:`[lang_code] X [eos]`, where :obj:`lang_code` is source
+language id for source text and target language id for target text, with :obj:`X` being the source or target text
+respectively.
```
```suggestion
```

patrickvonplaten commented on docs/source/model_doc/m2m_100.rst:
```diff
@@ -0,0 +1,127 @@
+    with tokenizer.as_target_tokenizer():
+        labels = tokenizer(tgt_text, return_tensors="pt").input_ids
+
+    model(**model_inputs, labels=labels) # forward pass
```
(nit) think  `loss = model(**model_inputs, labels=labels).loss` is a bit nicer

patrickvonplaten commented on docs/source/model_doc/m2m_100.rst:
```diff
@@ -0,0 +1,127 @@
+
+- Generation
+
+    To generate using the M2M100 multilingual translation models, :obj:`eos_token_id` is used as the
```
```suggestion
    M2M100 uses the :obj:`eos_token_id` as the
```

patrickvonplaten commented on docs/source/model_doc/m2m_100.rst:
```diff
@@ -0,0 +1,127 @@
+- Generation
+
+    To generate using the M2M100 multilingual translation models, :obj:`eos_token_id` is used as the
+    :obj:`decoder_start_token_id` and the target language id is forced as the first generated token. To force the
```
```suggestion
    :obj:`decoder_start_token_id` for generation with the target language id being forced as the first generated token. To force the
```

patrickvonplaten commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -143,6 +139,10 @@
 from ..lxmert.modeling_lxmert import LxmertForPreTraining, LxmertForQuestionAnswering, LxmertModel
+
+# Add modeling imports here
+# Add modeling imports here
```
this should probably be cleaned up (also cc @LysandreJik this looks like an error in the templates to me)

patrickvonplaten commented on src/transformers/models/m2m_100/configuration_m2m_100.py:
```diff
@@ -0,0 +1,164 @@
+
+    def __init__(
+        self,
+        vocab_size=128112,
```
that's pretty huge

patrickvonplaten commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1248 @@
+]
+
+
+def shift_tokens_right(input_ids: torch.Tensor, pad_token_id: int, decoder_start_token_id: int):
```
could you add a `#Copied from `statement here from Bart maybe?

patrickvonplaten commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1248 @@
+
+def shift_tokens_right(input_ids: torch.Tensor, pad_token_id: int, decoder_start_token_id: int):
+    """
+    Shift input ids one token to the right, and wrap the last non pad token (the <LID> token) Note that MBart does not
```
MBart?

patrickvonplaten replied:
The comment does not seem to be correct

patrickvonplaten commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1248 @@
+    return prev_output_tokens
+
+
+def _make_causal_mask(input_ids_shape: torch.Size, dtype: torch.dtype, past_key_values_length: int = 0):
```
please add `#Copied from` statements

patrickvonplaten commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1248 @@
+    return mask[None, None, :, :].expand(bsz, 1, tgt_len, tgt_len + past_key_values_length)
+
+
+def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
```
please add `#Copied from` statements

patrickvonplaten commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1248 @@
+    return incremental_indices.long() + padding_idx
+
+
+class M2M100SinusoidalPositionalEmbedding(nn.Module):
```
please add `#Copied from` statements (if possible here not sure with the offset)

patrickvonplaten commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1248 @@
+        return position_ids.unsqueeze(0).expand(input_shape).contiguous()
+
+
+class M2M100Attention(nn.Module):
```
please add `#Copied from` statements (if possible here not sure with the offset)

patrickvonplaten reviewed (commented):
Great work @patil-suraj - could you add as many `#Copied from` statements as possible in the `modeling_...py` file and ping me again for review?

patil-suraj commented:
Sure, Patrick !

patil-suraj committed:
add copied from statements

patil-suraj committed:
fix tokenizer tests

patil-suraj committed:
apply review suggestions

patil-suraj committed:
fix urls

patil-suraj commented on docs/source/pretrained_models.rst:
```diff
@@ -365,6 +365,12 @@ For the full list, refer to `https://huggingface.co/models <https://huggingface.
+| M2M100             | ``facebook/m2m100_418M``                                   | | 24-layer, 1024-hidden, 16-heads, 418M parameters                                                                                    |
+|                    |                                                            | | multilingual machine translation model for 100 languages                                                                            |
+|                    +------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------+
+|                    | ``facebook/m2m100_1.2B``                                   | | 24-layer, 1024-hidden, 16-heads, 610M parameters                                                                                    |
```
still need to port 1.2B checkpoint, will update the layer info after porting

patil-suraj committed:
fix shift_tokens_right

patil-suraj commented on tests/test_modeling_m2m_100.py:
```diff
@@ -0,0 +1,372 @@
+class M2M100ModelIntegrationTests(unittest.TestCase):
+    @cached_property
+    def default_tokenizer(self):
+        return M2M100Tokenizer.from_pretrained("valhalla/m2m100_418M")
```
I Will change the model names, once I push the checkpoints to` facebook/`

sgugger commented on docs/source/model_doc/m2m_100.rst:
```diff
@@ -0,0 +1,125 @@
+
+.. code-block::
+
+    from transformers import M2M100ForConditionalGeneration, M2M100Tokenizer
+
+    hi_text = "जीवन एक चॉकलेट बॉक्स की तरह है।"
+    chinese_text = "生活就像一盒巧克力。"
+
+    model = M2M100ForConditionalGeneration.from_pretrained("facebook/m2m100_418M")
+    tokenizer = M2M100Tokenizer.from_pretrained("facebook/m2m100_418M")
+
+    # translate Hindi to French
+    tokenizer.src_lang = "hi"
+    encoded_hi = tokenizer(hi_text, return_tensors="pt")
+    generated_tokens = model.generate(**encoded_hi, forced_bos_token_id=tokenizer.get_lang_id("fr"))
+    tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)
+    # => "La vie est comme une boîte de chocolat."
```
Since there is a result here, this should use the doctest syntax (so >>> for each line of code, nothing for the return.

sgugger commented on docs/source/model_doc/m2m_100.rst:
```diff
@@ -0,0 +1,125 @@
+    tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)
+    # => "La vie est comme une boîte de chocolat."
+
+    # translate Chinese to English
+    tokenizer.src_lang = "ar_AR"
+    encoded_zh = tokenizer(chinese_text, return_tensors="pt")
+    generated_tokens = model.generate(**encoded_zh, forced_bos_token_id=tokenizer.get_lang_id("en"))
+    tokenizer.batch_decode(generated_tokens, skip_special_tokens=True)
+    # => "Life is like a box of chocolate."
```
Same here.

sgugger commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -21,8 +21,6 @@
 from ...configuration_utils import PretrainedConfig
 from ...file_utils import add_start_docstrings
 from ...utils import logging
-
-# Add modeling imports here
```
This should not be removed in this PR as it is used by the model templates to detect where the new model imports should be added.

sgugger commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -66,8 +64,6 @@
     CamembertForTokenClassification,
     CamembertModel,
 )
-
-# Add modeling imports here
```
This one can stay as it was added by mistake.

sgugger commented on src/transformers/models/m2m_100/configuration_m2m_100.py:
```diff
@@ -0,0 +1,164 @@
+    Args:
+        vocab_size (:obj:`int`, `optional`, defaults to 50265):
+            Vocabulary size of the M2M100 model. Defines the number of different tokens that can be represented by the
+            :obj:`inputs_ids` passed when calling :class:`~transformers.M2M100Model` or
+            :class:`~transformers.TFM2M100Model`.
```
No TF model
```suggestion
            :obj:`inputs_ids` passed when calling :class:`~transformers.M2M100Model`.
```

sgugger commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1312 @@
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
+
+    @property
+    def dummy_inputs(self):
+        pad_token = self.config.pad_token_id
+        input_ids = torch.tensor([[0, 6, 10, 4, 2], [0, 8, 12, 2, pad_token]], device=self.device)
+        dummy_inputs = {
+            "attention_mask": input_ids.ne(pad_token),
+            "input_ids": input_ids,
+        }
+        return dummy_inputs
```
Why is this here? PyTorch models don't require dummy inputs.

patrickvonplaten replied:
Actually, I don't know why there were there in the first place in Bart & I just left them. But I think we should do a clean-up where we remove all those `dummy_inputs` from all Bart models. I'm ok for leaving it in this PR for now and remove all of them in a follow-up PR.

sgugger commented on src/transformers/models/m2m_100/tokenization_m2m_100.py:
```diff
@@ -0,0 +1,327 @@
+    Examples::
+
+        >>> from transformers import M2M100Tokenizer
+        >>> tokenizer = M2M100Tokenizer.from_pretrained("facebook/m2m100_418M, src_lang="en_XX", tgt_lang="ro_RO")
```
Love the set at init! But it seems from what's below that the language codes are not the same as mBART.
```suggestion
        >>> tokenizer = M2M100Tokenizer.from_pretrained("facebook/m2m100_418M, src_lang="en", tgt_lang="ro")
```

sgugger reviewed (approved):
Great addition! Out of curiosity, what is missing to get a fast tokenizer like for mBART?

LysandreJik commented on src/transformers/models/m2m_100/configuration_m2m_100.py:
```diff
@@ -0,0 +1,164 @@
+
+        Example::
+
+        >>> from transformers import M2M100Model, M2M100Config
+
+        >>> # Initializing a M2M100 m2m_100_418M style configuration
+        >>> configuration = M2M100Config()
+
+        >>> # Initializing a model from the m2m_100_418M style configuration
+        >>> model = M2M100Model(configuration)
+
+        >>> # Accessing the model configuration
+        >>> configuration = model.config
```
This needs to be indented otherwise it doesn't render well in the docs

LysandreJik reviewed (approved):
This looks great! Would it be hard to implement the fast tokenizer as well?

patrickvonplaten commented on src/transformers/models/m2m_100/configuration_m2m_100.py:
```diff
@@ -0,0 +1,164 @@
+        init_std=0.02,
+        decoder_start_token_id=2,
+        scale_embedding=True,
+        gradient_checkpointing=False,
```
Docs for gradient_checkpointing are missing (are they missing in the templates as well?) -> could you maybe add them there in this case as well @patil-suraj ?

patrickvonplaten commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1312 @@
+    are ignored. This is modified from fairseq's `utils.make_positions`.
+
+    Args:
+        x: torch.Tensor x:
```
can you improve the docstring here? there is not input `x`

patrickvonplaten commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1312 @@
+class M2M100SinusoidalPositionalEmbedding(nn.Module):
+    """This module produces sinusoidal positional embeddings of any length."""
+
+    def __init__(self, num_positions: int, embedding_dim: int, padding_idx: Optional[int] = None):
```
let's remove the padding index -> there should be no padding index in the positional embeddings. See [this](https://github.com/huggingface/transformers/pull/10200) discussion.

patrickvonplaten commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1312 @@
+        self.register_buffer("_float_tensor", torch.FloatTensor(1))
+
+    @staticmethod
+    def get_embedding(num_embeddings: int, embedding_dim: int, padding_idx: Optional[int] = None):
```
also let's remove the padding_idx logic here -> there shouldn't be one for positions

patrickvonplaten commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1312 @@
+        >>> model_inputs = tokenizer(text_to_translate, return_tensors='pt')
+
+        >>> # translate to French
+        >>> gen_tokens = model.generate( **model_inputs, forced_bos_token_id=tok.get_lang_id("fr"))
```
really like the new "forced_bos_token_id" design - nice!

patrickvonplaten commented on src/transformers/models/m2m_100/modeling_m2m_100.py:
```diff
@@ -0,0 +1,1312 @@
+        self.embed_positions = M2M100SinusoidalPositionalEmbedding(
+            config.max_position_embeddings,
+            embed_dim,
+            self.padding_idx,
```
think we sholdu remove the padding idx

patrickvonplaten reviewed (approved):
Great job! Looks very clean & thanks for adding all those copied from statements!

Left a couple of smaller improvement suggestions:

1) One doc was missing
2) Think we should not continue this practice of giving the padding_idx a special behavior in the postiional embeddings
3) We should move all checkpoints to `facebook` before merging

patil-suraj committed:
apply review suggestions

patil-suraj committed:
fix

patil-suraj committed:
fix doc

patil-suraj committed:
add lang code to id

patil-suraj committed:
remove unused function

patil-suraj committed:
update checkpoint names

patil-suraj committed:
fix copy

patil-suraj committed:
fix tokenizer

patil-suraj committed:
fix checkpoint names

patil-suraj committed:
fix merge issue

patil-suraj committed:
style

patil-suraj commented:
I’ve addressed all the review comments, and all the slow/fast tests are now passing.

I didn’t add fast tokenizer because `M2M100` is `sentencepiece`  based tokenizer, but it uses `sentencepiece`  for just tokenizing and then uses a vocab file to convert the tokens to ids and ids to tokens. So our current `SpmConverter` doesn’t work for this. I’ll try to add fast tokenizer in a follow-up PR.

Merging!

Muennighoff commented:
> I’ve addressed all the review comments, and all the slow/fast tests are now passing.
> 
> I didn’t add fast tokenizer because `M2M100` is `sentencepiece` based tokenizer, but it uses `sentencepiece` for just tokenizing and then uses a vocab file to convert the tokens to ids and ids to tokens. So our current `SpmConverter` doesn’t work for this. I’ll try to add fast tokenizer in a follow-up PR.
> 
> Merging!

Hey, I was wondering if there's any progress on a Fast Tokenizer for M2M or if any help can be needed?
Thanks :)
